### ISSUE TITLE
Seq Lab: Illumina Runsheet performance and bugs

### ISSUE THREAD

**Issue by tim-snyder-wustl** [GTAC-MGI/GTAC-ESP-LIMS#403]
From Matt:

>What I am seeing seems to be random with what the new script generated. If the pool is just libraries, it is expanded and the pool is just listed on a line by itself as well. If the pool contains libraries and pool, then the pools are expanded, but the libraries are not there. However, that is not always the case, another pool has some pools expanded and others aren’t there… but always no libraries in those cases of pool and libraries.

- Runsheet should show *only* libraries, not dilutions.

**Comment by tim-snyder-wustl** [comment]
[227GLVLT4-illumina-runsheet.ts.csv](https://github.com/GTAC-MGI/GTAC-ESP-LIMS/files/15152312/227GLVLT4-illumina-runsheet.ts.csv)

**Comment by tim-snyder-wustl** [comment]
[illumina-setup-esp-export.csv](https://github.com/GTAC-MGI/GTAC-ESP-LIMS/files/15152524/illumina-setup-esp-export.csv)

![Image](https://github.com/GTAC-MGI/GTAC-ESP-LIMS/assets/91081406/71beaf1d-aaed-47b5-8e1b-5e5f682dd651)

**Comment by tim-snyder-wustl** [comment]
Testing on:
https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/70f3915c-37f4-4fc8-9d32-c43cdbfed0a0

with Robert Z's script.

throws error:

```
Traceback (most recent call last):
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 258, in <module>
    prot_sample_to_libs = resolve_libraries(protocol)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 150, in resolve_libraries
    protocol_samp_to_libs, lib_to_pool = _resolve_libs_from_pools_or_libs(all_pools_or_libs, protocol_samp_to_pool, pool_uuids)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 104, in _resolve_libs_from_pools_or_libs
    raise AssertionError(f"Expected pool uuid {pool} to be in the uuid_to_libs mapping, but it wasn't. Available keys: {list(uuid_to_libs.keys())}")
AssertionError: Expected pool uuid 0006755e-a896-439e-a9b8-83b69d6ab800 to be in the uuid_to_libs mapping, but it wasn't. Available keys: ['0229b1d5-3685-4425-8bba-7cd9694d205c', '13401ec2-7e1a-4538-ab38-7f0a07ffa67b', '21459f7f-5292-482c-9a7c-4e26b214fa17', '22b986b2-8b23-4e8a-b908-906b6b3fb577', '318db7b2-ea37-4bf5-8348-d9f59eefd7b2', '366e16a8-95bf-48b6-ab45-23d6f00b842e', '3df19f40-3d3d-4378-a5c7-c9ac0c212b29', '475ed4e9-f97e-4982-8108-e8bf512bc461', '5131a35f-9f3d-4b1c-8349-c902abe4634b', '56dbe878-b663-47b7-ba8c-fee97694a47e', '7d91f3bd-a861-4e3d-ad17-ef040717822f', '7e800943-8796-463e-aa62-cca74a255134', '8032988e-a663-402f-82ec-6c1d93b7bbd1', '928c76a3-5867-41c7-8e9c-2ac01cbb5426', '98abd0d5-c646-4d77-8b7c-83696c9d2eae', 'a98c45d0-af7f-4f12-906c-0bed2da78c8a', 'b7551a82-33b3-4ec0-9ae5-f3ba6acba1c9', 'bb8a35d5-62b9-4e55-b36c-e29fe9b23c14', 'bc6a6eaa-faaa-4e51-bb9a-98297811e140', 'c67b3d4a-d31a-4d61-978f-eceb815ba021', 'd6301a4a-435b-4ba3-b6af-1357d981f69b', 'e90eb0a2-ffd7-4480-b9e5-d32127b69ba2', 'ea40ab2a-5a57-4b67-ab6c-f42e79c4ad88', 'ea546742-4497-47aa-a837-da2111104271', 'ecb00d47-280c-45d6-a90e-a02d14cef52c', 'f2374f4f-5f24-48dc-83e1-4d94374c2388', 'f636a995-c72b-44c0-8d4c-39066137c702']
```

**Comment by tim-snyder-wustl** [comment]
Also testing: 

https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/d75583e4-0d7f-4bbe-80d2-5fc575c2e6a5
https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/0ff0e944-0793-4d93-8dc7-b87e24234d18
https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/9ac9694f-8734-4a6e-bb7f-3dcce8b5f9bf

**Comment by tim-snyder-wustl** [comment]
Feedback from Matt re TS script version:

>Looks like 227GMVLT4 is accurate.
 For 22C7KKLT4, looks like all the libraries that were pooled without dilutions are missing from the new sheet. 11 total.
 22C7J2LT4, lanes 1-5 are OK, but many samples missing from lanes 6,7 and 8. I can’t really see why those are missing.
 In the end, both sheets should always contain the same number of rows.

**Comment by tim-snyder-wustl** [comment]
New error on Prod:
https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/a91adfb0-52e3-4109-ba0a-fd88429a71c9

```
Traceback (most recent call last):
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 266, in <module>
    update_entities_flowcell_data(protocol, prot_sample_to_libs)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 204, in update_entities_flowcell_data
    update_data(ancestor, flowcell_data, row['Flowcell Type'], flowcell_cycles)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 176, in update_data
    sample_specs.append(sample.push(dry=True))
  File "/opt/l7esp/data/extensions/client/lib/python3.9/site-packages/esp/models/__base__.py", line 2287, in push
    payload[key] = self.__push_format__[key](self)
  File "/opt/l7esp/data/extensions/client/lib/python3.9/site-packages/esp/models/sample.py", line 651, in <lambda>
    __push_format__ = {"resource_vals": lambda x: x._push_variables(x)}
  File "/opt/l7esp/data/extensions/client/lib/python3.9/site-packages/esp/models/sample.py", line 683, in _push_variables
    vals = raw(obj.resource_vals)
  File "/opt/l7esp/data/extensions/client/lib/python3.9/site-packages/esp/models/__base__.py", line 1227, in __getattr__
    raise AttributeError("{} object has no attribute {}".format(self.__class__.__name__, name))
AttributeError: Entity object has no attribute resource_vals
```

**Comment by tim-snyder-wustl** [comment]
New error on Prod:
https://gtac-mgi-lims.wustl.edu/#/lims/worksheet/d75583e4-0d7f-4bbe-80d2-5fc575c2e6a5

```
Traceback (most recent call last):
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 265, in <module>
    prot_sample_to_libs = resolve_libraries(protocol)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 157, in resolve_libraries
    protocol_samp_to_libs, lib_to_pool = _resolve_libs_from_pools_or_libs(all_pools_or_libs, protocol_samp_to_pool, pool_uuids)
  File "/opt/l7esp/data/content/pipelines/scripts/generate_illumina_runsheet.py", line 111, in _resolve_libs_from_pools_or_libs
    raise AssertionError(f"Expected pool uuid {pool} to be in the uuid_to_libs mapping, but it wasn't. Available keys: {list(uuid_to_libs.keys())}")
AssertionError: Expected pool uuid 9c94c1ea-1641-4ae1-b4df-9797dea4b2c7 to be in the uuid_to_libs mapping, but it wasn't. Available keys: ['00887953-18d3-445a-a773-bb5c2bce9c08', '09f27bc5-78a3-4d04-b59d-a294ccd9947f', '1325bdef-e757-4416-8976-119e6273d724', '28f2d5b2-19ea-4088-9581-7096dafae6c5', '497d1b95-9b8d-4af7-946b-fb4c0e83942f', '536ceaa8-7d70-474b-8743-39267f4c8ac7', '787f4304-f686-47f3-8270-1270e6e59647', '7c6fe840-556f-4840-b1fc-9e5696541d66', '8144613c-129e-4fb2-8aaf-6544bf87184e', '82fd6aa3-622e-47f3-a9e9-d16163b491f0', '9692ad7c-6bf1-4ab1-95f0-65b7b9a4b7fa', 'a056981b-c1ba-44e5-912e-cfb9eaf81970', 'b1489c00-018d-47dd-8d86-373677044a97', 'c943cc70-b506-46bd-800a-30b15ca4c792', 'd0390b2d-e00b-4649-9e0a-7de3a352a21c', 'd3f40b24-5334-4f90-af2b-620203991daa', 'd54db3d5-a044-48e6-af47-b69d6b8a917f', 'dbd40954-7fe0-438a-af5c-317d13d1e122', 'de478fea-1234-45b3-8e2a-1197844448e2', 'f59c9b0c-1227-43f8-a1c0-c5cc66c49ac4', 'f6e3f8a4-7809-4f12-a747-d501698ed75e', 'ff579541-e8c0-450d-a487-0712507ee1f5']
```
